### PR TITLE
Allow toggling touchscreen mode at runtime

### PR DESCRIPTION
**Pull request by okias**
#### What

Allow runtime switching between desktop and touchscreen mode.

#### How

This PR replaces hardcoded `#ifdef`s with setting (`enable_touch`).

#### Resolves:
 - Fixes (but just partly) https://github.com/minetest/minetest/issues/10062
 - Fixes https://github.com/minetest/minetest/issues/12785

#### Problem solved by this:

Ability to distribute one binary for touchscreen enabled and classic (desktops and laptops without touchscreens).

https://floss.social/@okias/111417351641760633

People love to play Minetest on Linux phones. Some of these phones are convergent, so they work as desktop when connected to USB-C.

Also we have Linux tablets and convertible PCs.

## How to test

In the settings appear new checkbox for the touchscreen. You can check it or uncheck it and see whats going to happen.

#### TODO
 - At some point would be lovely to have touchscreen getting auto-detected.

#### Depends
 - on 5.9 (currently `master` branch), for code applicable on 5.8, use https://github.com/okias/minetest/tree/linux-touch-5.8
 - https://github.com/minetest/minetest/pull/14117
 - https://github.com/minetest/minetest/pull/14146
 - https://github.com/minetest/minetest/pull/13282

### CODE REVIEW

File: builtin/settingtypes.txt [hunk 1]
```diff
@@ -148,6 +148,11 @@ invert_hotbar_mouse_wheel (Hotbar: Invert mouse wheel direction) bool false
 
 [*Touchscreen]
 
+#    Enables the touch screen.
+#    If disabled, touchscreen wouldn't be usable in-game
+#    Changing this setting requires a restart.
+enable_touch (Enable Touchscreen) bool true
```
Comment (rubenwardy) [inline builtin/settingtypes.txt:154]:
```suggestion
#    Enables touchscreen controls, allowing you to play the
#    game with a touchscreen
#
#    Changing this setting requires a restart.
enable_touch_controls (Enable Touchscreen Controls) bool true
```

**Comment by rubenwardy** [comment]
For clarify, it would be good to change this to `enable_touch_controls` or `touchscreen_controls`.   TouchScreenGUI should also be renamed to TouchScreenControls

File: builtin/settingtypes.txt (hunk 1 above)
Comment (grorp) [reply builtin/settingtypes.txt:154]: It shouldn't actually require a restart, though.

Comment (rubenwardy) [reply builtin/settingtypes.txt:154]: Yeah I agree

Comment (okias) [reply builtin/settingtypes.txt:154]: What you propose is the optimal solution, thou I think at this moment it's not only the controls which this option enable, but it alters between two modes: mouse X touchscreen. When touchscreen (controls) is enabled, mouse stops working. So far this MR is maximally simple replacement of hard coded compile-only option. It doesn't bring additional improvements regarding to being able use mouse control while using touchscreen.

Comment (rubenwardy) [reply builtin/settingtypes.txt:154]: When touchscreen controls are disabled, you can still use the touchscreen to interact with GUIs as it emulates mouse movements

Comment (rubenwardy) [reply builtin/settingtypes.txt:154]: Ahh, even if the touchscreen emulates mouse presses. It won't handle gestures. So this setting does effect the gui too

File: src/client/game.cpp [hunk 2]
```diff
@@ -1273,9 +1259,8 @@ void Game::shutdown()
 	if (formspec)
 		formspec->quitMenu();
 
-#ifdef HAVE_TOUCHSCREENGUI
-	g_touchscreengui->hide();
-#endif
+	if (g_settings->getBool("enable_touch"))
+		g_touchscreengui->hide();
```
Comment (grorp) [inline src/client/game.cpp:1263]: `enable_touch` can change anytime, so only hiding `g_touchscreengui` if `enable_touch` is true doesn't suffice.

For example, set `enable_touch = true` and put this into a mod:

```lua
minetest.register_on_joinplayer(function()
    minetest.after(1, function()
        minetest.settings:set("enable_touch", "false")
        minetest.request_shutdown()
    end)
end)
```

Result: The touchscreen controls stay visible on the shutdown screen.

There are probably many more problems like this with this PR. That's why I repeat my suggestion:
 1. only set `g_touchscreengui` to something other than null if `enable_touch` is true
 2. and then check for `g_touchscreengui` instead of `enable_touch`

We can then make in-game switching work properly using `SettingsChangedCallback`.

**Review by grorp — CHANGES_REQUESTED** [review]

File: src/client/game.cpp (hunk 2 above)
Comment (okias) [reply src/client/game.cpp:1323]: Does that mean you agree to decide in the client launcher if we will have touchscreen support or not and then sticking to it?

Comment (grorp) [reply src/client/game.cpp:1323]: > Does that mean you agree to decide in the client launcher if we will have touchscreen support or not and then sticking to it?

Yes, I want "touchscreen controls enablement" to be decided here, in `clientlauncher.cpp`:

https://github.com/minetest/minetest/blob/32e492837cbf286aeb91b4b63ecf3c890c71a1bc/src/client/clientlauncher.cpp#L259-L272

This code runs each time you start a game from the mainmenu.

To make switching in-game possible (without going back to the mainmenu first), we can use a `SettingsChangedCallback`. This doesn't necessarily need to happen in this PR.

File: src/script/cpp_api/s_base.cpp [hunk 3]
```diff
@@ -153,13 +153,6 @@ ScriptApiBase::ScriptApiBase(ScriptingType type):
 	lua_pushstring(m_luastack, porting::getPlatformName());
 	lua_setglobal(m_luastack, "PLATFORM");
 
-#ifdef HAVE_TOUCHSCREENGUI
```
Comment (okias) [inline src/script/cpp_api/s_base.cpp:156]: Should I go back for LUA to something similar too?

File: src/client/clientlauncher.cpp [hunk 4]
```diff
@@ -256,10 +256,8 @@ bool ClientLauncher::run(GameStartData &start_data, const Settings &cmd_args)
 			m_rendering_engine->get_video_driver()->setTextureCreationFlag(
 					video::ETCF_CREATE_MIP_MAPS, g_settings->getBool("mip_map"));
 
-#ifdef HAVE_TOUCHSCREENGUI
 			receiver->m_touchscreengui = new TouchScreenGUI(m_rendering_engine->get_raw_device(), receiver);
 			g_touchscreengui = receiver->m_touchscreengui;
```
[the remark this answers is not in the document]
Comment (okias) [reply src/client/clientlauncher.cpp]: resolved

File: src/defaultsettings.cpp [hunk 5]
```diff
@@ -192,7 +192,11 @@ void set_default_settings()
 	settings->setDefault("screen_h", "600");
 	settings->setDefault("window_maximized", "false");
 	settings->setDefault("autosave_screensize", "true");
+#ifdef ENABLE_TOUCH
+	settings->setDefault("fullscreen", "true");
+#else
 	settings->setDefault("fullscreen", "false");
+#endif
```
[the remark this answers is not in the document]
Comment (okias) [reply src/defaultsettings.cpp:199]: Because on Linux phones, otherwise you miss part of the screen (kinda as on Android). It's not optimal for the desktop thou.

File: src/gui/touchscreengui.cpp [hunk 6]
```diff
@@ -106,7 +106,10 @@ static EKEY_CODE id_to_keycode(touch_gui_button_id id)
 		default:
 			break;
 	}
-	assert(!key.empty());
+	if (key.empty()) {
+		assert(!key.empty());
```
Comment (okias) [inline src/gui/touchscreengui.cpp:110]: IMHO the actions shouldn't be translated to keys. We'll have a problem when a user switches from the keyboard where he may deassign keys and then back to the touchscreen, where everything should work just fine.

File: src/defaultsettings.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (grorp) [reply src/defaultsettings.cpp:199]: Hmm, I guess nobody is going to set `ENABLE_TOUCH=TRUE` unless they're compiling for a (Android/Linux/whatever) phone or tablet, right? Then it's probably fine.

File: src/script/cpp_api/s_base.cpp (hunk 3 above)
Comment (grorp) [reply src/script/cpp_api/s_base.cpp:156]: I'm not sure what you mean. Your current approach looks good.

If you mean going back to the "check for `g_touchscreengui`" approach: That's not possible since we're talking about mainmenu Lua code, where `g_touchscreengui` is always null.

File: src/gui/touchscreengui.cpp (hunk 6 above)
Comment (grorp) [reply src/gui/touchscreengui.cpp:110]: Yes, I agree, but that would require enough refactoring to consider it out of scope for this PR.

```cpp
	if (key.empty()) {
		assert(!key.empty());
```

What? ;)

Comment (grorp) [reply src/gui/touchscreengui.cpp:110]: My approach regarding the unassigned key problem for this PR would be:

1. Only set a "rangeselect" key in `defaultsettings.cpp` if `ENABLE_TOUCH` is defined, i.e. if `enable_touch` is going to be enabled by default.

2. In `touchscreengui.cpp`, only show a button in the first place if the respective key is defined.

**Review by grorp — CHANGES_REQUESTED** [review]
Thank you for the changes. With the new approach, this already looks much better.

Here are some comments, I haven't looked very closely yet.

File: builtin/mainmenu/content/dlg_contentstore.lua [hunk 7]
```diff
@@ -1175,8 +1175,8 @@ end
 
 function store.handle_events(event)
 	if event == "DialogShow" then
-		-- On mobile, don't show the "MINETEST" header behind the dialog.
-		mm_game_theme.set_engine(TOUCHSCREEN_GUI)
+		-- On touchscreen, Don't show the "MINETEST" header behind the dialog.
```
Comment (grorp) [inline builtin/mainmenu/content/dlg_contentstore.lua:1178]:
```suggestion
		-- On touchscreen, don't show the "MINETEST" header behind the dialog.
```

File: src/clientdynamicinfo.h [hunk 8]
```diff
@@ -68,7 +68,7 @@ struct ClientDynamicInfo
 #ifndef SERVER
 	static v2f32 calculateMaxFSSize(v2u32 render_target_size, f32 gui_scaling) {
 		f32 factor =
-#ifdef HAVE_TOUCHSCREENGUI
+#ifdef __ANDROID__
```
Comment (grorp) [inline src/clientdynamicinfo.h:71]: This should check for `enable_touch` directly because this code also does:

https://github.com/minetest/minetest/blob/5ac13ed7f0dd70a82631cbc9119792a3c59d08ed/src/gui/guiFormSpecMenu.cpp#L3285-L3292

File: src/gui/guiPasswordChange.cpp [hunk 9]
```diff
@@ -66,11 +63,8 @@ void GUIPasswordChange::regenerateGui(v2u32 screensize)
 	/*
 		Calculate new sizes and positions
 	*/
-#ifdef HAVE_TOUCHSCREENGUI
 	const float s = m_gui_scale * RenderingEngine::getDisplayDensity() / 2;
```
Comment (grorp) [inline src/gui/guiPasswordChange.cpp:66]: There are two problems with this:

1. It's a weird hack, `m_gui_scale` already takes the display density into account.

2. It causes the "change password" dialog to become unusably small for some display density values.

```suggestion
	const float s = m_gui_scale;
```

**Review by grorp — CHANGES_REQUESTED** [review]
Looked a bit closer :)

File: src/client/game.cpp [hunk 10]
```diff
@@ -4416,9 +4383,15 @@ void Game::readSettings()
 
 void Game::showDeathFormspec()
 {
+	std::string size_tag;
+	if (g_touchscreengui)
+		size_tag = "size[11,5.5]";
+	else
+		size_tag = "size[11,5.5,true]"; // Fixed size on desktop
```
Comment (grorp) [inline src/client/game.cpp:4390]: duplicated code for `size_tag`, should be moved to a common function

File: src/client/game.cpp [hunk 11]
```diff
@@ -4440,61 +4413,65 @@ void Game::showDeathFormspec()
...
-		"- double tap (outside):\n"
-		" --> close\n"
-		"- touch stack, touch slot:\n"
-		" --> move stack\n"
-		"- touch&drag, tap 2nd finger\n"
-		" --> place single item to slot\n"
+	std::string control_text, size_tag;
+
+	if (g_touchscreengui) {
+		size_tag = "size[11,5.5]";
```
Comment (grorp) [inline src/client/game.cpp:4419]: duplicated code for `size_tag`, should be moved to a common function

File: builtin/settingtypes.txt (hunk 1 above)
Comment (grorp) [reply builtin/settingtypes.txt:154]: > `Enable Touchscreen`

This should be `Enable touchscreen`. Most setting names are in sentence case (there are some exceptions, but we don't have to make it worse).

> `Enables the touch screen.`

I'd write `Enables touchscreen mode.` instead. Also, looking at `settingtypes.txt`, we usually write "touchscreen" rather than "touch screen".

> `If disabled, touchscreen wouldn't be usable in-game`

I'd write something like `Allows you to play the game with a touchscreen.` instead (copied from rubenwardy).

> `Changing this setting requires a restart.`

This is plain wrong, it doesn't.

--------

So to sum up, I suggest this:

```suggestion
#    Enables touchscreen mode, allowing you to play the game with a touchscreen.
enable_touch (Enable touchscreen) bool true
```

**Review by grorp — CHANGES_REQUESTED** [review]

**Review by grorp — CHANGES_REQUESTED** [review]
And ... I have another comment. There's a leftover `HAVE_TOUCHSCREENGUI` here:

https://github.com/minetest/minetest/blob/5ac13ed7f0dd70a82631cbc9119792a3c59d08ed/src/clientdynamicinfo.h#L53-L57

It should be replaced by a check for `g_touchscreengui`.

---

It would be nice if you would apply my suggestions and any other changes without force-pushing (when possible). That would make reviewing easier.

File: src/gui/touchscreengui.cpp [hunk 12]
```diff
@@ -107,6 +107,9 @@ static EKEY_CODE id_to_keycode(touch_gui_button_id id)
 			break;
 	}
 	assert(!key.empty());
+	if (g_settings->get("keymap_" + key) == "")
+		return KEY_NONE;
```
Comment (grorp) [inline src/gui/touchscreengui.cpp:111]: Apparently `KEY_NONE` is mapped to a key on some systems:

```cpp
        KEY_NONE             = 0xFF,  // usually no key mapping, but some laptops use it for fn key
```

https://github.com/minetest/irrlicht/blob/03dd8b0336d041726dbec5e47dc8bfcbcf858767/include/Keycodes.h#L180

It would be better to use `KEY_KEY_CODES_COUNT` to mean "no key".

File: src/gui/touchscreengui.cpp [hunk 13]
```diff
@@ -107,6 +107,9 @@ static EKEY_CODE id_to_keycode(touch_gui_button_id id)
...
+
 	return keyname_to_keycode(g_settings->get("keymap_" + key).c_str());
```
Comment (grorp) [inline src/gui/touchscreengui.cpp:113]: This still throws an `UnknownKeycode` exception if the key name is invalid.

A possible solution:

```patch
diff --git a/src/client/keycode.cpp b/src/client/keycode.cpp
index 1fcd15e01b..14360b1a1c 100644
--- a/src/client/keycode.cpp
+++ b/src/client/keycode.cpp
@@ -18,7 +18,6 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 */
 
 #include "keycode.h"
-#include "exceptions.h"
 #include "settings.h"
 #include "log.h"
 #include "debug.h"
@@ -26,13 +25,6 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "util/string.h"
 #include "util/basic_macros.h"
 
-class UnknownKeycode : public BaseException
-{
-public:
-	UnknownKeycode(const char *s) :
-		BaseException(s) {};
-};
-
 struct table_key {
 	const char *Name;
 	irr::EKEY_CODE Key;
diff --git a/src/client/keycode.h b/src/client/keycode.h
index 7036705d16..6bc44bb1d4 100644
--- a/src/client/keycode.h
+++ b/src/client/keycode.h
@@ -19,11 +19,19 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 
 #pragma once
 
+#include "exceptions.h"
 #include "irrlichttypes.h"
 #include "Keycodes.h"
 #include <IEventReceiver.h>
 #include <string>
 
+class UnknownKeycode : public BaseException
+{
+public:
+	UnknownKeycode(const char *s) :
+		BaseException(s) {};
+};
+
 /* A key press, consisting of either an Irrlicht keycode
    or an actual char */
 
diff --git a/src/gui/touchscreengui.cpp b/src/gui/touchscreengui.cpp
index eb73008879..b7f034cda2 100644
--- a/src/gui/touchscreengui.cpp
+++ b/src/gui/touchscreengui.cpp
@@ -107,10 +107,18 @@ static EKEY_CODE id_to_keycode(touch_gui_button_id id)
 			break;
 	}
 	assert(!key.empty());
-	if (g_settings->get("keymap_" + key) == "")
-		return KEY_NONE;
 
-	return keyname_to_keycode(g_settings->get("keymap_" + key).c_str());
+	std::string resolved = g_settings->get("keymap_" + key);
+	EKEY_CODE code;
+	try {
+		code = keyname_to_keycode(resolved.c_str());
+	} catch (UnknownKeycode &e) {
+		code = KEY_KEY_CODES_COUNT;
+		if (!resolved.empty())
+			warningstream << "TouchScreenGUI: Unknown key '" << resolved
+					<< "' for '" << key << "', ignoring." << std::endl;
+	}
+	return code;
 }
 
 static void load_button_texture(const button_info *btn, const std::string &path,
@@ -534,7 +542,7 @@ void TouchScreenGUI::init(ISimpleTextureSource *tsrc)
 	};
 	std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
 	for (const auto &pair : buttons) {
-		if (id_to_keycode(pair.first) == KEY_NONE)
+		if (id_to_keycode(pair.first) == KEY_KEY_CODES_COUNT)
 			continue;
 
 		const wchar_t* wide = converter.from_bytes(pair.second).c_str();
```

**Review by grorp — CHANGES_REQUESTED** [review]

File: src/defaultsettings.cpp [hunk 14]
```diff
@@ -90,12 +95,7 @@ void set_default_settings()
 	settings->setDefault("keymap_cmd_local", ".");
 	settings->setDefault("keymap_minimap", "KEY_KEY_V");
 	settings->setDefault("keymap_console", "KEY_F10");
-#if HAVE_TOUCHSCREENGUI
-	// See https://github.com/minetest/minetest/issues/12792
 	settings->setDefault("keymap_rangeselect", "KEY_KEY_R");
-#else
-	settings->setDefault("keymap_rangeselect", "");
-#endif
```
[the remark this answers is not in the document]
Comment (grorp) [reply src/defaultsettings.cpp:98]: Repeating my suggestion from https://github.com/minetest/minetest/pull/14075#discussion_r1438169158:

> Only set a "rangeselect" key in `defaultsettings.cpp` if `ENABLE_TOUCH` is defined, i.e. if `enable_touch` is going to be enabled by default.

**Review by grorp — CHANGES_REQUESTED** [review]

File: src/gui/touchscreengui.cpp (hunk 12 above)
Comment (okias) [reply src/gui/touchscreengui.cpp:111]: I would suggest option `KEY_NONE` since it's kinda more descriptive? What do you think?

File: src/clientdynamicinfo.h [hunk 15]
```diff
@@ -50,11 +50,7 @@ struct ClientDynamicInfo
 		f32 hud_scaling = g_settings->getFloat("hud_scaling", 0.5f, 20.0f);
 		f32 real_gui_scaling = gui_scaling * density;
 		f32 real_hud_scaling = hud_scaling * density;
-#ifdef HAVE_TOUCHSCREENGUI
-		bool touch_controls = true;
-#else
-		bool touch_controls = false;
-#endif
+		bool touch_controls = g_settings->getBool("enable_touch");
```
[the remark this answers is not in the document]
Comment (okias) [reply src/clientdynamicinfo.h:53]: right, just had to add the `gui/touchscreengui.h` header file

File: src/gui/touchscreengui.cpp (hunk 13 above)
Comment (okias) [reply src/gui/touchscreengui.cpp:113]: Look good. Just moved `code` definition into beginning of the block

**Comment by okias** [comment]
@SmallJoker as a author of the original intent to run Minetest on PP CCing you too (and thank you for the initial work!)

File: src/gui/touchscreengui.cpp (hunk 13 above)
Comment (okias) [reply src/gui/touchscreengui.cpp]: Could we print warning even when var is empty? So people can have a clue why a TS button is missing?

Comment (grorp) [reply src/gui/touchscreengui.cpp]: > Could we print warning even when var is empty? So people can have a clue why a TS button is missing?  
> edit: I would drop the `if (!resolved.empty())` condition if you wouldn't mind

Makes sense. You could make the message even more helpful by replacing "ignoring" with "hiding button".

File: src/gui/touchscreengui.cpp (hunk 12 above)
Comment (grorp) [reply src/gui/touchscreengui.cpp]: Well, using `KEY_NONE` to mean "no key" would introduce a bug: If someone were to assign `KEY_NONE` to one of the TouchScreenGUI buttons, the button would be hidden. So I'd say we can't use it.

Also, we're already using `KEY_KEY_CODES_COUNT` to mean "no key" in `src/client/keycode.cpp`:

https://github.com/minetest/minetest/blob/2c44620e5e344ca2b4acaaaa5408502332ca5aff/src/client/keycode.cpp#L314-L320

**Comment by grorp** [comment]
Here's a patch to fix a small bug in the settings menu: https://github.com/grorp/minetest/commit/74d130f0807c1a1291e48311b2bb9d8524971b4b

**Comment by rubenwardy** [comment]
> Here's a patch to fix a small bug in the settings menu: https://github.com/grorp/minetest/commit/74d130f0807c1a1291e48311b2bb9d8524971b4b

This is a hack. Instead, you should remove "touchscreen" as a setting requirement and then make all touchscreen settings rely on "enable_touch".  The "keyboard_mouse" should then be set to `PLATFORM ~= "Android"`

**Comment by okias** [comment]
> For clarity, it would be good to change this to `enable_touch_controls` or `touchscreen_controls`. TouchScreenGUI should also be renamed to TouchScreenControls, and `touchscreen_gui` to `touchscreen_controls`

This sounds very reasonable, would you be okay if I set it after merging this PR? I'm trying to keep scope as small as possible to push this, as code changes often and it needs rebases :)

**Review by grorp — CHANGES_REQUESTED** [review]
- The "Controls" button (formerly "Change Keys") in the pause menu is currently hidden on Android. I think it should instead be hidden if the touch controls are active. Reasons:
   - Changing keybinds in-game may break the touch controls.
   - If Android players disable touch controls, they should be able to rebind keys in-game just like other players.

- Disabling touch controls on Android makes the game unplayable even if you have a mouse and a keyboard connected because Irrlicht's CIrrDeviceAndroid doesn't implement mouse input. This will hopefully be fixed by switching the Android builds to CIrrDeviceSDL.

   Until that happens, the `enable_touch` setting should be behind `Requires: !android`.

**Comment by Zughy** [comment]
@okias Whoops with those edits :D I was trying to link the issues, but since it's a partial fix, I've unlinked the former

**Comment by okias** [comment]
> * The "Controls" button (formerly "Change Keys") in the pause menu is currently hidden on Android. I think it should instead be hidden if the touch controls are active. Reasons:
>       
>       * Changing keybinds in-game may break the touch controls.
>       * If Android players disable touch controls, they should be able to rebind keys in-game just like other players.

I implemented the change, but in general I hope it'll be temporary measure, in future I would love to see key+touch supported, so keyboard settings could stay.

> 
>     * Disabling touch controls on Android makes the game unplayable even if you have a mouse and a keyboard connected because Irrlicht's CIrrDeviceAndroid doesn't implement mouse input. This will hopefully be fixed by switching the Android builds to CIrrDeviceSDL.
>       Until that happens, the `enable_touch` setting should be behind `Requires: !android`.

Done :)

**Comment by okias** [comment]
I quickly re-iterated and applied suggestions and implemented proposed changes. I wish I could invest more time, but until the end of the month, I'm deeply busy. Hope it looks good, any further comments are welcome and I'll happily update the PR.

**Comment by grorp** [comment]
> I implemented the change, but in general I hope it'll be temporary measure, in future I would love to see key+touch supported, so keyboard settings could stay.

You're right. I've reverted 0ddb30229cb7fbce991b640904803242e0107d59 and added the "changing keybinds in-game breaks TouchScreenGUI" bug to my to-do list.

I've also pushed another commit to avoid another round-trip.

**Review by grorp — APPROVED** [review]
This PR looks good now. It seems to work on my Linux desktop computer with a touchscreen attached, and it doesn't seem to break anything on Android.

`m_touch_simulate_aux1` is buggy, but that's already the case on 5.8.0 and not this PR's fault.

Thank you very much for your work on this PR, @okias!

In the future, CIrrDeviceAndroid should be replaced by CIrrDeviceSDL (for working mouse input) so that we can make use of this feature on Android.

**Comment by okias** [comment]
Re-tested, latest Minetest master + this MR; Irrlicht master + https://github.com/minetest/irrlicht/pull/278 .

IMHO Works well.

**Comment by okias** [comment]
@rubenwardy pretty please, one review for me (I think I'll have to keep rebasing it all the times otherwise)

**Comment by rubenwardy** [comment]
Hey, just tested this out on my Android phone and I'm unable to dig nodes. I press and hold but nothing happens. I cloned the `git merge-base` of this PR (2ef080a51b5982e4e2ce17953f2a9cc62edf13fa), and I can dig nodes. I made sure that Irrlicht is up to date.

**Comment by okias** [comment]
@rubenwardy do I understand right the MR works correctly when based on https://github.com/minetest/minetest/commit/2ef080a51b5982e4e2ce17953f2a9cc62edf13fa but digging not working when based on `master`?

**Comment by rubenwardy** [comment]
Digging didn't work for me on this MR (https://github.com/minetest/minetest/commit/842ca79aacb661b0720e85dd0659723bed59d187) but did work in the minetest master branch (https://github.com/minetest/minetest/commit/2ef080a51b5982e4e2ce17953f2a9cc62edf13fa)

**Comment by okias** [comment]
@rubenwardy found it, it was 404a063fdfbbddcba32d22dd1c34dd73922bc12a adding another check. We need to merge it rather sooner than later, it keeps breaking every few commits.

File: src/client/game.cpp [hunk 16]
```diff
@@ -2620,8 +2620,8 @@ void Game::updateCameraDirection(CameraOrientation *cam, float dtime)
 	Since Minetest has its own code to synthesize mouse events from touch events,
 	this results in duplicated input. To avoid that, we don't enable relative
 	mouse mode if we're in touchscreen mode. */
-	if (cur_control)
-		cur_control->setRelativeMode(!g_touchscreengui && !isMenuActive());
+	if (!g_touchscreengui && cur_control)
+		cur_control->setRelativeMode(!isMenuActive());
```
Comment (grorp) [inline src/client/game.cpp:2624]: This change is wrong, it will prevent disabling relative mouse mode if we implement a way to toggle touchscreen mode in-game. Why did you do it?

Comment (okias) [reply src/client/game.cpp]: sorry about that, it was the debugging process of non-working digging on Android and I kept the change. Removed.

**Comment by okias** [comment]
I have prepared basic auto-detection MR for Linux, which will set sane defaults regarding to touchscreen if knows the device form factor (tablet/handheld/phone).

**Comment by okias** [comment]
Tested on:
 - Android (works as expected)
 - Linux
   - TS enabled (phone)
   - TS disabled (laptop)
 - Windows (laptop w/ touchscreen)
   - TS disabled, works as expected
   - TS enabled, works same as on Linux (just touchpad/mouse digging doesn't work in TS mode)

File: src/client/game.cpp [hunk 17]
```diff
@@ -4548,9 +4515,9 @@ void Game::showPauseMenu()
 		<< strgettext("Exit to Menu") << "]";
```
Comment (sfan5) [inline src/client/game.cpp:4515]: the controls menu above here is still gated by `__ANDROID__`

File: src/CMakeLists.txt [hunk 18]
```diff
@@ -109,9 +109,10 @@ if(BUILD_CLIENT AND ENABLE_SOUND)
 	endif()
 endif()
 
-option(ENABLE_TOUCH "Enable Touchscreen support" FALSE)
+option(ENABLE_TOUCH "Enable touchscreen by default" FALSE)
```
Comment (sfan5) [inline src/CMakeLists.txt:112]: IMO we should just get rid of this option
or is there an use case for it?

File: builtin/mainmenu/settings/dlg_settings.lua [hunk 19]
```diff
@@ -641,11 +641,20 @@ local function buttonhandler(this, fields)
 		local value = core.is_yes(fields.show_advanced)
 		core.settings:set_bool("show_advanced", value)
 		write_settings_early()
+	end
+
+	if fields.enable_touch ~= nil then
```
Comment (sfan5) [inline builtin/mainmenu/settings/dlg_settings.lua:646]: where is the field this code handles defined?

File: src/client/game.cpp [hunk 20]
```diff
@@ -1076,9 +1070,7 @@ Game::Game() :
 
 	readSettings();
 
-#ifdef HAVE_TOUCHSCREENGUI
-	m_cache_hold_aux1 = false;	// This is initialised properly later
-#endif
+	m_touch_simulate_aux1 = false;	// This is initialised properly later
```
Comment (sfan5) [inline src/client/game.cpp:1073]: note: you could just move the default value to the definition point

File: src/client/game.cpp [hunk 21]
```diff
@@ -4548,9 +4515,9 @@ void Game::showPauseMenu()
 		<< strgettext("Exit to Menu") << "]";
 	os		<< "button_exit[4," << (ypos++) << ";3,0.5;btn_exit_os;"
 		<< strgettext("Exit to OS")   << "]";
-#ifdef HAVE_TOUCHSCREENGUI
+	if (g_touchscreengui) {
```
Comment (sfan5) [inline src/client/game.cpp:4518]:
```suggestion
	if (!control_text.empty()) {
```

File: src/gui/touchscreengui.cpp [hunk 22]
```diff
@@ -110,7 +111,15 @@ static EKEY_CODE id_to_keycode(touch_gui_button_id id)
 			break;
 	}
 	assert(!key.empty());
-	return keyname_to_keycode(g_settings->get("keymap_" + key).c_str());
+	std::string resolved = g_settings->get("keymap_" + key);
+	try {
+		code = keyname_to_keycode(resolved.c_str());
+	} catch (UnknownKeycode &e) {
+		code = KEY_KEY_CODES_COUNT;
```
Comment (sfan5) [inline src/gui/touchscreengui.cpp:118]: I think `KEY_UNKNOWN` would fit better

File: src/gui/touchscreengui.cpp [hunk 23]
```diff
@@ -523,13 +532,23 @@ void TouchScreenGUI::init(ISimpleTextureSource *tsrc)
...
 			AHBB_Dir_Right_Left, 3.0f);
 
-	m_settings_bar.addButton(fly_id, L"fly", "fly_btn.png");
-	m_settings_bar.addButton(noclip_id, L"noclip", "noclip_btn.png");
-	m_settings_bar.addButton(fast_id, L"fast", "fast_btn.png");
-	m_settings_bar.addButton(debug_id, L"debug", "debug_btn.png");
-	m_settings_bar.addButton(camera_id, L"camera", "camera_btn.png");
-	m_settings_bar.addButton(range_id, L"rangeview", "rangeview_btn.png");
-	m_settings_bar.addButton(minimap_id, L"minimap", "minimap_btn.png");
+	std::map<touch_gui_button_id, std::string> settings_bar_buttons {
```
Comment (sfan5) [inline src/gui/touchscreengui.cpp:535]: please make these `const static`

File: src/main.cpp [hunk 24]
```diff
@@ -50,9 +50,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "gui/guiEngine.h"
 #include "gui/mainmenumanager.h"
 #endif
-#ifdef HAVE_TOUCHSCREENGUI
-	#include "gui/touchscreengui.h"
-#endif
+#include "gui/touchscreengui.h"
```
Comment (sfan5) [inline src/main.cpp:53]: why is this even included here? I didn't see any usage from a quick check.

File: src/gui/touchscreengui.cpp (hunk 22 above)
Comment (grorp) [reply src/gui/touchscreengui.cpp:118]: oh lol, we discussed `KEY_NONE` (can't be used because it's a real key) vs. `KEY_KEY_CODES_COUNT`, but looks like both of us didn't see `KEY_UNKNOWN`

https://github.com/minetest/minetest/pull/14075#discussion_r1439275487

File: src/client/game.cpp (hunk 17 above)
Comment (grorp) [reply src/client/game.cpp:4515]: hmm, the controls / "change keys" menu could be useful even with `enable_touch=true` if you play with touchscreen (for camera movement) and keyboard (for everything else)

not sure why you'd do that though. we've already discussed this somewhere.

File: src/CMakeLists.txt (hunk 18 above)
Comment (grorp) [reply src/CMakeLists.txt:112]: the Android build sets it
mobile Linux distributions might want to set it, but it will be obsolete if we somehow implement proper auto-detection

Comment (sfan5) [reply src/CMakeLists.txt:112]: android is a solved issue: we have an ifdef in defaultsettings.cpp

Comment (okias) [reply src/CMakeLists.txt:112]: needed until we get auto-detection. I have this in followup MR.

File: src/client/game.cpp (hunk 17 above)
Comment (okias) [reply src/client/game.cpp:4515]: I would prefer to not touch it at this moment.

File: builtin/mainmenu/settings/dlg_settings.lua (hunk 19 above)
Comment (okias) [reply builtin/mainmenu/settings/dlg_settings.lua:646]: I'm afraid I don't understand the question.

Comment (srifqi) [reply builtin/mainmenu/settings/dlg_settings.lua:646]: I can not find it either.

File: src/gui/guiPasswordChange.cpp [hunk 25]
```diff
@@ -24,10 +24,7 @@ OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 #include <IGUIButton.h>
 #include <IGUIStaticText.h>
 #include <IGUIFont.h>
-
-#ifdef HAVE_TOUCHSCREENGUI
-	#include "client/renderingengine.h"
-#endif
+#include "client/renderingengine.h"
```
Comment (srifqi) [inline src/gui/guiPasswordChange.cpp:27]: This include is not used.

**Review by srifqi — CHANGES_REQUESTED** [review]
Please also remove `TOUCHSCREEN_GUI` from `.luacheckrc` file.

File: builtin/mainmenu/settings/dlg_settings.lua (hunk 19 above)
Comment (sfan5) [reply builtin/mainmenu/settings/dlg_settings.lua:646]: The contents of `fields` are set according to the state of the formspec and/or buttons pressed.
If there is no formspec field named "enable_touch" defined then this key can never show up in the table.
I couldn't find one, hence the question.

Comment (grorp) [reply builtin/mainmenu/settings/dlg_settings.lua:646]: The field is auto-generated from `settingtypes.txt`. That defines the key.

**Comment by okias** [comment]
Done

**Review by srifqi — APPROVED** [review]

**Review by grorp — APPROVED** [review]
Still looks good

File: builtin/mainmenu/settings/dlg_settings.lua (hunk 19 above)
Comment (andrewwardmt) [reply builtin/mainmenu/settings/dlg_settings.lua:646]: This is misleading placement, given that it's next to show_advanced but the actual checkbox is defined as a component based on the setting. I'd recommend adding a comment at least, or maybe moving this code to `make.bool` although maybe that violates separation of responsibiities

Comment (rubenwardy) [reply builtin/mainmenu/settings/dlg_settings.lua:646]: This is misleading placement, given that it's next to show_advanced but the actual checkbox is defined as a component based on the setting. I'd recommend adding a comment at least, or maybe moving this code to make.bool although maybe that violates separation of responsibiities

**Review by rubenwardy — APPROVED** [review]
Code looks good

**Comment by okias** [comment]
Follow up: https://github.com/minetest/minetest/pull/14400
